521 allow suppressing buildtime slot

HaraldKi opened the pull request:
- ran all `tox -e ...` commands, no (new) failures, the `1 skipped, 2 xfailed` when running the tests is the same as on master.
- added two unit tests to cover the new `--buildtime=None` feature but also check that previous behaviour still works.
- for easier local coverage inspection I added a coverage run to produce an HTML report to `tox.ini`.

Addresses issue #521.

HaraldKi committed:
Changes footer from plain template file to subclass of TemplateElement

Though it reads the same template file, footer.html, as before.

Command line option may now also be --buildtime=None to avoid showing the
buildtime string in the template shown with a transparent renderer for buildtime
inside the footer.

HaraldKi committed:
Add test to check --buildtime and the respective renderer in Footer

Also adds a tox section to get an HTML coverage report.

tristanlatr commented on pydoctor/driver.py:
```diff
@@ -44,11 +44,14 @@ def get_system(options: model.Options) -> model.System:
         try:
-            system.buildtime = datetime.datetime.strptime(
-                options.buildtime, BUILDTIME_FORMAT)
+            if options.buildtime == "None":
```
```suggestion
            if options.buildtime.lower() == "none":
```

HaraldKi replied:
I considered the `lower()` and then thought that pythoneers should be quite used to writing `None` all over the place. 

Of course it is not python but command line syntax. Maybe we should not confuse the two and use `--buildtime=no` instead?

What do you prefer?

tristanlatr reviewed (commented):
Hello @HaraldKi and many thanks for your Pull Request; it looks great!
I only have a little suggestion that is to move the ' at ' inside the template file like before. 
Also adding a changelog entry in the README would be appreciated.

Many thanks,

tristanlatr commented on pydoctor/templatewriter/pages/__init__.py:
```diff
@@ -241,7 +265,7 @@ def slot_map(self) -> Dict[str, "Flattenable"]:
             project=project_tag,
             pydoctor_version=__version__,
-            buildtime=system.buildtime.strftime("%Y-%m-%d %H:%M:%S"),
+            buildtime="" if system.buildtime is None else system.buildtime.strftime("%Y-%m-%d %H:%M:%S")
```
Let's remove it since people using custom footer.html templates will get a warning when updating to the next pydoctor version (because of the template version mismatch). So this will drag them to downloading the latest version of the template file.

tristanlatr commented on pydoctor/templatewriter/pages/__init__.py:
```diff
@@ -200,6 +201,29 @@ def title(self, request: IRequest, tag: Tag) -> str:
+        """Renders the build time as provided on the command line with a default of now."""
+        if self._buildtime is None:
+            return ""
+        ## FIXME: should we have the format as an option?
```
What about this: 

``<t:transparent fmt=" at %Y-%m-%d %H:%M:%S" t:render="buildtime"></t:transparent>``

This way, the format string is ensured to be a string, while being customizable by templates.

HaraldKi committed:
Changes how the format is used and how to suppress build time rendering

Rather than appending the build time to the transparent tag's content, we solely
use its fmt= attribute to generate the build time. The tag's content is ignored.

To suppress built time rendering any false valued string can be used, not None.q

tristanlatr committed:
Minor simplifications

tristanlatr committed:
Simplify

tristanlatr committed:
Add changelog entry

tristanlatr committed:
Remove tox env

tristanlatr committed:
Test all HTML files

tristanlatr committed:
Merge branch 'master' into 521_allow_suppressing_buildtime_slot

tristanlatr committed:
Fix syntax

tristanlatr committed:
Remove unused import

tristanlatr committed:
Add missing footer and javascript include to the all-documents/html template

tristanlatr committed:
Bump version of template file all-documents.html
